dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java

```diff
@@ -0,0 +1,127 @@
+ * Responsible for creating a map of gene to external db ids
+ */
+@RequiredArgsConstructor
+public final class ExternalReader {
+
+  /**
+   * Dependencies
+   */
+  private final XrefReader nameReader;
+  private final GeneReader geneReader;
+  private final TranslationReader translationReader;
+
+  /**
+   * State
+   */
+  @Getter
+  private final Map<String, ObjectNode> externalIds = new HashMap<>();
+
+  public ExternalReader read() {
+    val geneIdMap = geneReader.getGeneIdMap();
+    val transToGeneMap = translationReader.getTranslationToGene();
+
+    BaseReader.read(OBJECT_XREF_URI, line -> {
+      if ("Gene".equals(line[2])) {
+        String geneId = geneIdMap.get(line[1]);
+
+        ObjectNode externalDbs;
+        if (externalIds.containsKey(geneId)) {
+          externalDbs = externalIds.get(geneId);
+        } else {
+          externalDbs = DEFAULT.createObjectNode();
+          externalDbs.put("entrez_gene", DEFAULT.createArrayNode());
+          externalDbs.put("hgnc", DEFAULT.createArrayNode());
+          externalDbs.put("omim_gene", DEFAULT.createArrayNode());
+          externalDbs.put("uniprotkb_swissprot", DEFAULT.createArrayNode());
+          externalIds.put(geneId, externalDbs);
+        }
+
+        String xrefId = line[3];
+        if (nameReader.getEntrezMap().containsKey(xrefId)) {
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("entrez_gene");
+          arrayNode.add(nameReader.getEntrezMap().get(xrefId));
+
+        } else if (nameReader.getHgncMap().containsKey(xrefId)) {
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("hgnc");
+          arrayNode.add(nameReader.getHgncMap().get(xrefId));
+
+        } else if (nameReader.getMimMap().containsKey(xrefId)) {
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("omim_gene");
+          arrayNode.add(nameReader.getMimMap().get(xrefId));
+        }
+
+      } else if ("Translation".equals(line[2])) {
+        // Uniprot Ids are for proteins, which means we match them to translations and eventually work up to a gene/
+
+        String geneId = geneIdMap.get(transToGeneMap.get(line[1]));
+
+        ObjectNode externalDbs;
+        if (externalIds.containsKey(geneId)) {
+          externalDbs = externalIds.get(geneId);
+        } else {
+          externalDbs = DEFAULT.createObjectNode();
+          externalDbs.put("entrez_gene", DEFAULT.createArrayNode());
+          externalDbs.put("hgnc", DEFAULT.createArrayNode());
+          externalDbs.put("omim_gene", DEFAULT.createArrayNode());
+          externalDbs.put("uniprotkb_swissprot", DEFAULT.createArrayNode());
+          externalIds.put(geneId, externalDbs);
+        }
+
+        String xrefId = line[3];
+        if (nameReader.getUniprotMap().containsKey(xrefId)) {
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("uniprotkb_swissprot");
+          String uniprot = nameReader.getUniprotMap().get(xrefId);
+
+          Iterator<JsonNode> iter = arrayNode.elements();
```

Comment:
Like here: `hasUniprot(arrayNode)`? Actually, I think `ArrayNode` has methods for this.